crates/sui-types/src/effects/test_effects_builder.rs

```diff
@@ -0,0 +1,140 @@
+        }
+    }
+
+    pub fn with_status(mut self, status: ExecutionStatus) -> Self {
+        self.status = Some(status);
+        self
+    }
+
+    pub fn with_shared_input_versions(
+        mut self,
+        versions: BTreeMap<ObjectID, SequenceNumber>,
+    ) -> Self {
+        assert!(self.shared_input_versions.is_empty());
+        self.shared_input_versions = versions;
+        self
+    }
+
+    pub fn build(self) -> TransactionEffects {
+        let status = self.status.unwrap_or_else(|| ExecutionStatus::Success);
+        // TODO: This does not yet support deleted shared objects.
+        let shared_objects = self
+            .shared_input_versions
+            .iter()
+            .map(|(id, version)| SharedInput::Existing((*id, *version, ObjectDigest::MIN)))
+            .collect();
+        let executed_epoch = 0;
+        let lamport_version = SequenceNumber::lamport_increment(
+            self.transaction
+                .transaction_data()
+                .input_objects()
+                .unwrap()
+                .iter()
+                .filter_map(|kind| kind.version())
+                .chain(
+                    self.transaction
+                        .transaction_data()
+                        .receiving_objects()
+                        .iter()
+                        .map(|oref| oref.1),
+                )
+                .chain(self.shared_input_versions.values().copied()),
+        );
+        let sender = self.transaction.transaction_data().sender();
+        // TODO: Include receiving objects in the object changes as well.
+        let changed_objects = self
+            .transaction
+            .transaction_data()
+            .input_objects()
+            .unwrap()
+            .iter()
+            .filter_map(|kind| match kind {
+                InputObjectKind::ImmOrOwnedMoveObject(oref) => Some((
+                    oref.0,
+                    EffectsObjectChange {
+                        input_state: ObjectIn::Exist((
+                            (oref.1, oref.2),
+                            Owner::AddressOwner(sender),
+                        )),
+                        output_state: ObjectOut::ObjectWrite((
+                            // Digest must change with a mutation.
+                            ObjectDigest::MAX,
+                            Owner::AddressOwner(sender),
+                        )),
+                        id_operation: IDOperation::None,
+                    },
+                )),
+                InputObjectKind::MovePackage(_) => None,
+                InputObjectKind::SharedMoveObject {
+                    id,
+                    initial_shared_version,
+                    mutable,
+                } => mutable.then_some((
+                    *id,
+                    EffectsObjectChange {
+                        input_state: ObjectIn::Exist((
+                            (
+                                *self.shared_input_versions.get(id).unwrap(),
```

Comment:
Nit: looks like this shared_input_versions input is not optional. Consider adding a comment in TestEffectsBuilder saying that shared_input_versions is required for txn with shared objects. Or here maybe it could use a default value if it is not provided?